tests/aws-cpp-sdk-s3-unit-tests/S3UnitTests.cpp

```diff
@@ -0,0 +1,93 @@
+#include <gtest/gtest.h>
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/testing/mocks/http/MockHttpClient.h>
+#include <aws/testing/AwsTestHelpers.h>
+
+using namespace Aws;
+using namespace Aws::Auth;
+using namespace Aws::Http;
+using namespace Aws::S3;
+using namespace Aws::S3::Model;
+
+const char* ALLOCATION_TAG = "S3UnitTest";
+
+class S3UnitTest : public testing::Test {
+protected:
+  void SetUp() override {
+    _mockClientFactory = Aws::MakeShared<MockHttpClientFactory>(ALLOCATION_TAG);
+    _mockHttpClient = Aws::MakeShared<MockHttpClient>(ALLOCATION_TAG);
+    _mockClientFactory->SetClient(_mockHttpClient);
+    HttpOptions httpOptions;
+    httpOptions.httpClientFactory_create_fn = [this]() -> std::shared_ptr<HttpClientFactory> {
+      return _mockClientFactory;
+    };
+    _options.httpOptions = httpOptions;
+    InitAPI(_options);
+    AWSCredentials credentials{"mock", "credentials"};
+    _s3Client = Aws::MakeUnique<S3Client>("ALLOCATION_TAG", credentials);
+  }
+
+  void TearDown() override {
+    ShutdownAPI(_options);
+  }
+
+  SDKOptions _options;
+  std::shared_ptr<MockHttpClient> _mockHttpClient;
+  std::shared_ptr<MockHttpClientFactory> _mockClientFactory;
+  UniquePtr<S3Client> _s3Client;
+};
+
+
+TEST_F(S3UnitTest, S3UriLeadingDots) {
+  auto putObjectRequest = PutObjectRequest()
+      .WithBucket("bluerev")
+      .WithKey("../BoredInBristol");
+
+  std::shared_ptr<IOStream> body = Aws::MakeShared<StringStream>(ALLOCATION_TAG,
+    "Bored in Bristol, always waiting",
+    std::ios_base::in | std::ios_base::binary);
+
+  putObjectRequest.SetBody(body);
+
+  //We have to mock requset because it is used to create the return body, it actually isnt used.
+  auto mockRequest = Aws::MakeShared<Standard::StandardHttpRequest>(ALLOCATION_TAG, "mockuri", HttpMethod::HTTP_GET);
+  mockRequest->SetResponseStreamFactory([]() -> IOStream* {
+    IOStream* stream = Aws::New<StringStream>(ALLOCATION_TAG, "response-string", std::ios_base::in | std::ios_base::binary);
```

Comment:
"MEMORY LEAK" -- you shout

it is deleted when `mockResponse` is deleted in turn calling the destructor on its [bodyStream](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/http/standard/StandardHttpResponse.h#L72), which in turn calls [delete on the underlying stream](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/utils/stream/ResponseStream.cpp#L73-L87).